### PR TITLE
Feat/support custom domain for cognito user email notifications

### LINKED ISSUE #149: email deliverability config
- SPF, DKIM, DMARC

### PR DESCRIPTION
**Pull request by landonshumway-ia**
### Requirements List
- NOTE: This update requires an email address to receive SES email bounce and complaint notifications. It must be added to the cdk.context.json for each deployment environment under the following field:

```
"environments": {
    "<env_name>": {
      "notifications": {
        "ses_operations_support_email": "<operation team email address here>"
....
```

### Description List
- In order for the cognito user pool to send emails
from the custom domain, we need to set up an SES
identity along with necessary DNS records to verify
that the domain may be used to send emails.
This also sets up an SNS topic which will forward
all bounce or complaint notifications to an
operations team email address.
- The SES identity needs to be able to forward bounce
and complaint notifications for the domain, so that
an operations team can be made aware if a large spike
of message failures occurs for whatever reason. This
adds a required 'ses_feedback_forwarding_email' context
variable, which should be set to the email address of the
operations team that will manage these SES notifications.
- The custom domain name is an optional field. If it is not provided, the default cognito email settings will be used.

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- Code review

Closes #149

### CODE REVIEW

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: backend/compact-connect/stacks/persistent_stack/__init__.py [hunk 1]
```diff
@@ -68,8 +87,14 @@ def __init__(
             environment_name=environment_name,
             environment_context=environment_context,
             encryption_key=self.shared_encryption_key,
+            user_pool_email=user_pool_email_settings,
             removal_policy=removal_policy
         )
+        if self.hosted_zone:
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/persistent_stack/__init__.py:93]: helpful thanks!

File: backend/compact-connect/stacks/persistent_stack/user_email_notifications.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            master_key: IKey,
+            **kwargs
+    ):
+        super().__init__(scope, construct_id, **kwargs)
+
+        domain_name = hosted_zone.zone_name
+        operation_email = environment_context['notifications']['ses_operations_support_email']
+
+
+        self.email_feedback_topic = Topic(self, "FeedbackTopic",
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/persistent_stack/user_email_notifications.py:30]: Is this the formatting your linter uses? Not something I used to, but if that's what it enforces, all good.

File: backend/compact-connect/README.md
Comment (jlkravitz) [inline backend/compact-connect/README.md:1]: # General Comments
* This generally looks great to me. Thanks for the thorough comments and explanations of tricky configuration here.
* This seems hard to test but I always have to ask: Is there anything here we can easily test? e.g., DKIM, SPF, DMARC specification, etc. My instinct says not easily, but you're the expert!

**Review by jlkravitz — COMMENTED** [review]
looking good! a few nits/questions but can hopefully merge this today. (also, good to meet you, @landonshumway-ia!)

File: backend/compact-connect/README.md
Comment (landonshumway-ia) [reply backend/compact-connect/README.md:1]: Hi @jlkravitz. As far as testing the configuration, it's actually somewhat straightforward. Once you have added an email address for the "ses_operations_support_email" cdk.context.json field and deployed to your test env. The simplest way to verify is to first go the AWS SES console and create an SES verified email identity for your email address, See [Creating an email address identity](https://docs.aws.amazon.com/ses/latest/dg/creating-identities.html#verify-email-addresses-procedure). 

Once you verified your email address, go to the AWS Cognito console and find your user pool. From there, you have the option to create a new user using your verified email address, and select the option to send an email invite. Once you create the user, you should receive an email notification telling you your user was created, and you can verify that the FROM address is using your custom domain. The DMARC authentication will reject any emails from your domain that are not properly configured using SPF and DKIM, so if you get the email notification from cognito, you've verified that the authentication is working as expected. 

Now that I've walked through that, I guess it's not super straightforward, but it is verifiable. Let me know if you have any other questions about this.

File: backend/compact-connect/stacks/persistent_stack/user_email_notifications.py (hunk 2 above)
Comment (landonshumway-ia) [reply backend/compact-connect/stacks/persistent_stack/user_email_notifications.py:30]: Yes, if the function call causes the line length to be too long, my formatter puts the kwargs on separate lines. If there is a different formatting style that is preferred, let me know and I can adjust my IDE settings for the project.

File: backend/compact-connect/README.md [hunk 3]
```diff
@@ -131,6 +136,8 @@ that is done, perform the following steps to deploy the CI/CD pipeline into the
   the next step.
 - Create a new Route53 hosted zone for the domain name you plan to use for the app in each of the production AWS
   account and the test AWS account. See [About Route53 hosted zones](#about-route53-hosted-zones) below for more detail.
+- Request AWS to remove your account from the SES sandbox and wait for them to complete this request.
+  See [SES Sandbox](https://docs.aws.amazon.com/ses/latest/dg/request-production-access.html).
```
[the remark this answers is not in the document]
Comment (landonshumway-ia) [reply backend/compact-connect/README.md:140]: I've added a comment to the sandbox deploy section to walk through the option of requesting removal from 
the SES sandbox environment.

File: backend/compact-connect/README.md
Comment (jlkravitz) [reply backend/compact-connect/README.md:1]: Ha got it. I really meant testing in a python automated test suite, but this is also good :) Is this something we can add to the documentation?

File: backend/compact-connect/stacks/persistent_stack/user_email_notifications.py (hunk 2 above)
Comment (jlkravitz) [reply backend/compact-connect/stacks/persistent_stack/user_email_notifications.py:30]: Hmm strange. Is this consistent with the rules used by @jusdino? I guess I'm agnostic but would like us to be consistent. (Eventually, I hope we can commit our lint rules to the repo.)

Comment (jusdino) [reply backend/compact-connect/stacks/persistent_stack/user_email_notifications.py:30]: Our style rules are already in the repo [here](https://github.com/csg-org/CompactConnect/blob/development/backend/pylintrc) and this code does pass. For what it's worth, the [PEP8 guidelines](https://peps.python.org/pep-0008/#indentation) allow for a few different approaches to indentation (including this one), any of which may be best suited to a particular circumstance, so I don't know how more prescriptive we want to be. With these construct declarations, with function signatures of exactly two positional arguments followed by a list of keyword arguments, I've been fond of the aesthetic of a newline after the opening bracket, a line with the two positional arguments, then a new line for each keyword argument. Honestly, though, while I like the look of it, it is strictly the farthest departure from the PEP8 examples:
```python
topic = Topic(
    self, 'FeedbackTopic',
    display_name="Email Feedback Forwarding Topic",
    master_key=master_key,
    enforce_ssl=True
)
```

File: backend/compact-connect/README.md
Comment (landonshumway-ia) [reply backend/compact-connect/README.md:1]: Good call, I've added another section to the README walking through this verification process.

Comment (jlkravitz) [reply backend/compact-connect/README.md:1]: great!!

File: backend/compact-connect/stacks/persistent_stack/user_email_notifications.py (hunk 2 above)
Comment (jlkravitz) [reply backend/compact-connect/stacks/persistent_stack/user_email_notifications.py:30]: Ah honestly that example is what I thought the "standard" was :')

Given our discussion at today's retro let's table this and we can address globally when we get a linter set up.

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen this is good to merge!

**Pull request merged**